### ISSUE TITLE
Improve the accessibility of Code Radio app

### ISSUE THREAD

**Issue by huyenltnguyen** [freeCodeCamp/freeCodeCamp#42236]
**Affected page**

https://coderadio.freecodecamp.org/

**To Reproduce**

I made a quick scan using the [axe DevTools](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/) browser add-on and found that the page has some critical accessibility issues. 

We have 5 errors when the music is not played.

<img width="1680" alt="Screen Shot 2021-05-25 at 00 17 29" src="https://user-images.githubusercontent.com/25715018/119384616-2cd08200-bcef-11eb-8f5b-01a5ebb8c162.png">

And 9 when the music is playing (this has to do with the color contrast of the visualizer and the indicator).

<img width="1680" alt="Screen Shot 2021-05-25 at 00 23 13" src="https://user-images.githubusercontent.com/25715018/119384826-75883b00-bcef-11eb-9706-867479bb3022.png">

**To Resolve**

- Remove `aria-label` from the `canvas` element
- Add `aria-label` to the `button` element for song history 
- Add `aria-label` to the `select` element used for stream selection
- Remove `user-scalable=no` from the `meta` tag as well as set its `maximum-scale` to a value that is not less than 5. ([axe requires the value to not be less than 2](https://dequeuniversity.com/rules/axe/3.3/meta-viewport), but [Lighthouse sets the bar at 5](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/audits/accessibility/meta-viewport.js#L21-L22)) 
- Revisit the color contrast of the elements in question

**Comment by ish-u** [comment]
Hello,
I would like to help with this issue.

**Comment by gikf** [comment]
Hey @ish-u

issues are typically not assigned to specific person, other than long-time contributors, you can read more on https://contribute.freecodecamp.org/#/FAQ?id=how-can-i-get-an-issue-assigned-to-me. Issues with *help wanted* label are open for contributions for everybody.

Please make sure you read [our guidelines for contributing](https://contribute.freecodecamp.org/#/). Join us in [our chat room](https://chat.freecodecamp.org/channel/contributors) or [the forum](https://forum.freecodecamp.org/c/contributors/3) if you need help contributing, our moderators will guide you through this.

**Comment by ish-u** [comment]
Thank You for replying.
(Sorry i am new to open-source)

**Comment by ahmadabdolsaheb** [comment]
Also this is the repo for coderadio. https://github.com/freeCodeCamp/coderadio-client
I would recommend running it locally and skimming the source code before your first contribution.

**Comment by ish-u** [comment]
Hello, 
I need help regarding the last point - 'Revisit the color contrast of the elements in question"
Do I need to change the Colors ?
Also removing  `arai-label`  from Canvas component is not allowed by the Linter
\
![image](https://user-images.githubusercontent.com/51030453/121421087-6f4ebb80-c98b-11eb-9e22-cb34be2f51bb.png)

**Comment by ahmadabdolsaheb** [comment]
@ish-u, thanks for working on this. If you could list the elements in question, we could change their contrasts accordingly as the tool requires sign up and subscription plan.
The keyboard shortcuts button has lower opacity by design, but we could change that accordingly.

**Comment by ish-u** [comment]
Thank You for merging my PR.
 
On Google Chrome
  - while not Playing:
     -  Keyboard Shorcut 
  - while Playing:
     -  Keyboard Shorcut 
     -  Stream Selection

On Mozilla FireFox:
  - while not Playing:
     -  Keyboard Shorcut 
  - while Playing:
     -  Keyboard Shorcut 
     -  Stream Selection
     - CurrentSong Component 
        - title div
        - artist div
        - album div

**Comment by ahmadabdolsaheb** [comment]
Here is the design style guide: https://design-style-guide.freecodecamp.org/

And here is the colors that are used in the UI: https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/components/layouts/variables.css#L2-L30

feel free to replace the colors and create a pull request, I will go in the pr and change the colors if necessary.

**Comment by huyenltnguyen** [comment]
Hey team, sorry for not being able to respond sooner.

@ish-u Thank you for the fix! Don't worry, I can take a look at the canvas aria-label issue 🙂 

Regarding the color contrast, we don't have to address them all. axe DevTools, just like other audit tools, cannot be 100% accurate on the errors they flag as they can't determine if the color contrast actually needs to meet the requirement.

> On Google Chrome
> - while not Playing:
>   - Keyboard Shorcut
> - while Playing:
>   - Keyboard Shorcut
>   - Stream Selection
>
> On Mozilla FireFox:
> - while not Playing:
>   - Keyboard Shorcut
> - while Playing:
>   - Keyboard Shorcut
>   - Stream Selection
>   - CurrentSong Component
>     - title div
>     - artist div
>     - album div

I think the text color of song metadata is sufficient. It is `#ffffff` on `#0A0a23`, which yields a `19.43:1` in contrast radio. Same goes for the stream selection as I believe we don't have any custom CSS on it, so it looks the same as the native browser's `<select>`. 

I'm not on subscription plan either, so I'm not sure which element the tool refers to. Here are just a couple elements I could think of:
- Keyboard shortcut (as Anmol listed)
- Visualizer
- Progress indicator

We should address the contrast issue for (1) as it may not be distinguishable by users with low vision. If we think it would be distracting for the users, we can consider making it a popup and add a icon button to the footer to trigger it, similar to how the song history is displayed.

For the visualizer, I think we can leave it as is, since it is just a decorative component and does not communicate any information or require user's attention. From the "Graphic Objects" section of [this guide](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html):

> The term "graphical object" applies to stand-alone icons such as a print icon (with no text), and the important parts of a more complex diagram such as each line in a graph. For simple graphics such as single-color icons the entire image is a graphical object. Images made up of multiple lines, colors and shapes will be made of multiple graphical objects, some of which are required for understanding.
>
> Not every graphical object needs to contrast with its surroundings - only those that are required for a user to understand what the graphic is conveying.

I'm a bit concerned with the progress indicator, as it does provide information to the user, thus, needs to meet the contrast ratio requirement, especially when we don't have the song duration on the screen. However, since non-visual users cannot see the progress, we should still render the timer as text and use CSS to visually hide it if we don't want to show it on the screen. I'm not sure if the API returns this piece information for us, or we might need to do the calculation ourselves. We can split this into another issue and address it separately though.

**Comment by huyenltnguyen** [comment]
Huh, the `aria-label` in canvas is trickier than I thought.

According to the [HTML specs](https://www.w3.org/TR/html-aria/#dfn-any-role), `canvas` can take any ARIA attributes, so I think the this is a bug of the tool

<img width="805" alt="Screen Shot 2021-06-13 at 04 11 08" src="https://user-images.githubusercontent.com/25715018/121789118-80066800-cbfd-11eb-9422-c3e583460ece.png">

However, according to [Deque (the organization behind the Axe tool)](https://dequeuniversity.com/assets/pdf/module-nontext/module-nontext-checklist.pdf) and [an a11y expert](https://pauljadam.com/demos/canvas.html), the `role="img"` should be set to non-interactive `canvas`. But the `jsx-a11y` plugin is currently not allowing it https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/issues/648. 

There is already a PR for the eslint rule, and we should wait for it before making any change there.

**Comment by ish-u** [comment]
I also had the doubts regarding the `contrast` because the tool said something like "needs to be reviewed".
Maybe we can do something about the `Keyboard Shorcuts` like move it to the left side of Navigation Bar

**Comment by huyenltnguyen** [comment]
IMHO, having the keyboard shortcut box in the Footer component would be more appropriate, as the [`nav` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/nav) is usually used for important navigation links.

I was thinking about placing it on the right side of the Footer, both because it has been on the right side of the screen all along, and since it is less important than other elements, it should be the last item in the compositional flow. That said, I don't have a strong opinion on which side of the Footer it should be. 

I think we can make use of the [fa-keyboard-o](https://fontawesome.com/v4.7/icon/keyboard-o) btw 🙂

**Comment by ish-u** [comment]
Font Awesome icon would be better

**Comment by ahmadabdolsaheb** [comment]
Sure. adding it to the footer with that icon would make sense. 
That being said we will have another button that will toggle the embedded rocketchat.
I was wondering if we should add some info regarding reporting issues as well. 
That way a question mark icon or an info icon might be more appropriate.

**Comment by ish-u** [comment]
I can try adding the `Issue Report Button`. 
where should i place it on `navbar` on right or left?

**Comment by huyenltnguyen** [comment]
@ahmadabdolsaheb Would the chat toggle and the report issue button be separate items or they would be grouped in a menu? I guess grouping them would be better, as the footer would be a little cramped up otherwise.

**Comment by ahmadabdolsaheb** [comment]
How about this

[history] [----------------------------] [play/pause]  [volume] [get help(question mark)] [chat]

get help menu would have the keyboard shortcuts and 
report an issue( should go to the forum under the support category. )

**Comment by bbsmooth** [comment]
Are you still looking for help with accessibility enhancements for the radio? I'm currently preparing for the Web Accessibility Specialist certification offered by the IAAP and would be happy to help out.

**Comment by ahmadabdolsaheb** [comment]
Yes, please feel free to take a look.
Additionally, we would also appreciate taking a look at the editor in our new curriculum. Here is a link: https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-html-by-building-a-cat-photo-app/step-1

**Comment by Ajob41** [comment]
What is this for? Sorry I am little confused.

**Comment by bbsmooth** [comment]
@ahmadabdolsaheb I've started making accessibility updates to the radio and have already submitted three PRs.

**Comment by ahmadabdolsaheb** [comment]
@Ajob41, I agree that this issue's main focus has shifted. I will summarize and recreate another one shortly.
@bbsmooth thanks for the prs.

**Comment by Ajob41** [comment]
Sounds good.

**Comment by ahmadabdolsaheb** [comment]
thanks to @bbsmooth a lot of the accessibility issues have been resolved, so I will close this issue accordingly.
We can continue the discussion around the footer optimizations in a related issue if necessary.
